### PR TITLE
fix(Core/Game): Rest & Inn Improvements

### PR DESCRIPTION
**Pull request by Meltie2013**
## CHANGES PROPOSED:
-  Implemented missing Rest Flag Masks
-  Improved rest flag checks for Inns, Cities & Sanctuaries
-  Remove nasty hack fixes that just made no sense
-  Attempt to improve area trigger radius checks

## ISSUES ADDRESSED:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/3712

## TESTS PERFORMED:
- Watch the video here by @Maurowin: https://youtu.be/BwmdouFTv9Y

## HOW TO TEST THE CHANGES:
- Login a newly created character or pre-existing character (race doesn't matter)
- Head to any Inn, City or Sanctuary zone
- Check if player gets rest flag on entering the area
- Leave the zone and see if flag is updated

## KNOWN ISSUES AND TODO LIST:
- Radius around Inns are to large and require further updates are mentioned below in the feedback.

## Target branch(es):
- [x] Master

## Feedback
To fix radius issues around Inns, more updates are required on Position functions across the core. Math for it is incorrect and most of the time does nothing. Another PR will need to be opened for that separately. Changes made thus far are working as intended according to @Maurowin and the test they have done.

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by Maurowin** [comment]
https://youtu.be/BwmdouFTv9YSounds pretty good to me!
However, then remains to be confirmed, can the two factions take advantage of the rest in the hostels of the opposing faction?

If yes, it does not work.

The zone of rest is a little too wide and that it sees particularly in the inn of "Sentinel Hills" at the beginning and especially at the end of the video. I get the rest bonus while I'm 3 to 4 meters outside the hostel.

Translated with www.DeepL.com/Translator (free version)

**Comment by Meltie2013** [comment]
@Maurowin The wide radius stand point, it seems like a different issue. I can add a check to make sure the rest flag is removed when leaving the Inn.

**Comment by Maurowin** [comment]
@Meltie2013 ooh, so we should open a new issue ?

**Comment by Meltie2013** [comment]
@Maurowin It's a possibility till it can be confirmed.

**Comment by Maurowin** [comment]
@Meltie2013 concerning this point : can the two factions take advantage of the rest in the Inn of the opposing faction?
u know something ? 
To be confirmed but I think that both factions can benefit from it

**Comment by Meltie2013** [comment]
@Maurowin Alliance and Horde should only benefit from their own Inn's. From what I can remember back in the day of retail, this should be a big no.

**Comment by FALL1N1** [comment]
Inns are faction-based unless contested/neutral (e.g dalaran / shattrath / booty bay / gadgetzan...etc)

**Comment by Meltie2013** [comment]
@FALL1N1 That is also true, that been fixed in this PR as well. No checks existed for that.

**Comment by Maurowin** [comment]
The radius still to large and I have the feeling that the exit of the inns zone takes time to update. In the capitals there is no problem.
https://youtu.be/MK_7RnVZTjc

**Comment by Maurowin** [comment]
@FrancescoBorzi u can see my videos of my tests.

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -3023,20 +3032,22 @@ void Player::SetInWater(bool apply)
...
     if (!trigger || GetMapId() != trigger->map)
         return false;
 
-    if (trigger->radius > 0.f)
+    if (trigger->radius > 0)
     {
         // if we have radius check it
         float dist = GetDistance(trigger->x, trigger->y, trigger->z);
-        if (dist > trigger->radius)
+        if (dist > trigger->radius + delta)
```
Comment (BarbzYHOOL) [inline src/server/game/Entities/Player/Player.cpp:3044]: what the fuck is that???

so to fix it, "they" added artificial 5 meters???????? this is the "correct solution" ?? this is as dirty as before

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -1776,6 +1777,14 @@ void Player::Update(uint32 p_time)
     {
         if (p_time >= m_zoneUpdateTimer)
         {
+            // On zone update tick check if we are still in an inn if we are supposed to be in one
```
Comment (BarbzYHOOL) [inline src/server/game/Entities/Player/Player.cpp:1780]: really no better way can be found??

File: src/server/game/Entities/Player/Player.cpp [hunk 3]
```diff
@@ -7687,18 +7678,11 @@ void Player::UpdateArea(uint32 newArea)
...
-            RemoveByteFlag(UNIT_FIELD_BYTES_2, 1, UNIT_BYTE2_FLAG_FFA_PVP);
-    }
-    else if (!(areaFlags & AREA_FLAG_CAPITAL))
-    {
-        AreaTrigger const* atEntry = sObjectMgr->GetAreaTrigger(GetInnTriggerId());
-        if (!atEntry || !IsInAreaTriggerRadius(atEntry))
-            RemoveRestState();
-    }
+    uint32 const areaRestFlag = (GetTeamId() == TEAM_ALLIANCE) ? AREA_FLAG_REST_ZONE_ALLIANCE : AREA_FLAG_REST_ZONE_HORDE;
+    if (area && area->flags & areaRestFlag)
```
Comment (BarbzYHOOL) [inline src/server/game/Entities/Player/Player.cpp:7682]: can't you make use of the existing `area->IsInn(GetTeamId(true))` instead?

File: src/server/game/Entities/Player/Player.cpp [hunk 4]
```diff
@@ -7658,27 +7669,7 @@ void Player::UpdateArea(uint32 newArea)
...
-    bool isInn = area->IsInn(GetTeamId(true));
-    if (zone)
-    {
-        areaFlags |= zone->flags;
-        isSanctuary |= zone->IsSanctuary();
-        isInn |= zone->IsInn(GetTeamId(true));
-    }
-
-    // previously this was in UpdateZone (but after UpdateArea) so nothing will break
-    if (isSanctuary)    // in sanctuary
```
Comment (BarbzYHOOL) [inline src/server/game/Entities/Player/Player.cpp:7681]: can u eventually tell testers how to verify these things are still working somehow ingame after the rework?

**Comment by BarbzYHOOL** [comment]
@Maurowin can you test to set the worldserver.conf server type to FFAPVP, then open two clients. Your characters should have the "gurubashi arena pvp icon". Then you go with both inside an inn and you see if the icon disappears and if you can't attack each other.

Please

**Comment by BarbzYHOOL** [comment]
> @Meltie2013 ooh, so we should open a new issue ?

no it's not an issue, it's the inn radius in the DBC, it's blizzlike and it's just probably made to ease the calculation and the world building creation for the blizzard's devs.

File: src/server/game/Entities/Player/Player.cpp (hunk 2 above)
Comment (BarbzYHOOL) [reply src/server/game/Entities/Player/Player.cpp:1780]: ok rochet2 the boss enlightened me on why the client can't be trusted when exiting the inn, hence why there is prob no other efficient way to check when out of the inn. Client could never send the "not in inn anymore" and so players would cheat the resting system forever

so i guess no better way

**Comment by BarbzYHOOL** [comment]
> The radius still to large and I have the feeling that the exit of the inns zone takes time to update. In the capitals there is no problem.
> https://youtu.be/MK_7RnVZTjc

that's because when you leave a capital you exactly leave the zone. When you leave an inn, you're not leaving a zone per se so there is not a perfect trigger (hence why the +5 yards and also the tiny delay)

**Comment by Maurowin** [comment]
> 
> 
> @Maurowin can you test to set the worldserver.conf server type to FFAPVP, then open two clients. Your characters should have the "gurubashi arena pvp icon". Then you go with both inside an inn and you see if the icon disappears and if you can't attack each other.
> 
> Also can you test if logging out in capitals is instant?
> 
> Please

Will test this tomorrow

**Comment by Meltie2013** [comment]
@Maurowin I added additional check to update zone when player's position is updated. Make sure to pull this before additional testing is made.

File: src/server/game/Entities/Player/Player.cpp (hunk 4 above)
Comment (Meltie2013) [reply src/server/game/Entities/Player/Player.cpp:7681]: These checks are handled else where in the core, so they were removed.

File: src/server/game/Entities/Player/Player.cpp (hunk 2 above)
Comment (Meltie2013) [reply src/server/game/Entities/Player/Player.cpp:1782]: This is needed, else you would be back at square one of the rest flags not updating again.

File: src/server/game/Entities/Player/Player.cpp (hunk 3 above)
Comment (Meltie2013) [reply src/server/game/Entities/Player/Player.cpp:7693]: When `GetTeamId` is called, its looking for the team id.  Adding additional true or false really doesn't do anything but swapping the positions of true and false;

Better way of handling it.
```cpp
TeamId GetTeamId() const { return m_team == ALLIANCE ? TEAM_ALLIANCE : TEAM_HORDE; }
```

Comment (BarbzYHOOL) [reply src/server/game/Entities/Player/Player.cpp:7693]: i mean that "area->IsInn(GetTeamId(true))" is already pretty much:
```cpp
    uint32 const areaRestFlag = (GetTeamId() == TEAM_ALLIANCE) ? AREA_FLAG_REST_ZONE_ALLIANCE : AREA_FLAG_REST_ZONE_HORDE;
    if (area && area->flags & areaRestFlag)
```

**Comment by Maurowin** [comment]
There is no evolution and I even have the impression that it is necessary to go further when we leave the inn. 

The update " going out" must have a big problem.
https://youtu.be/crYCMO9YL9o

**Comment by BarbzYHOOL** [comment]
ok so we need to revert https://github.com/azerothcore/azerothcore-wotlk/pull/3780/commits/f3af4725ec970c06bdaaeb5942746a6b7f99a221 ??

**Comment by Maurowin** [comment]
The previous resquest sweater was slightly better.

Can we get inspiration from the code that manages the ffa mode when we go in and out of a hostel? Because this one works perfectly and the resquest should work exactly like it.

WoWScrnShot_112920_215008
WoWScrnShot_112920_215012

**Comment by Meltie2013** [comment]
@BarbzYHOOL This commit https://github.com/azerothcore/azerothcore-wotlk/commit/f3af4725ec970c06bdaaeb5942746a6b7f99a221 will be reverted.

**Comment by Maurowin** [comment]
It's better !

So like we talking in Discord, updateStatus dont work as expected.

Great job !

**Review by Helias — APPROVED** [review]

**Comment by BarbzYHOOL** [comment]
@Maurowin ok we gonna merge this, it's much better than before ingame (even though pixel perfect) and i talked with meltie and rochet, this seems like the best way to proceed and the most complete + a nice refactor of the ugly code we had

I put the "to be merged" label, if anyone doesn't agree, speak loudly!

**Review by Si1ker — APPROVED** [review]

**Pull request merged**